nucypher/utilities/metrics.py

```diff
@@ -1,23 +1,52 @@
 try:
-    from prometheus_client import Gauge, Enum, Counter, Info
+    from prometheus_client import Gauge, Enum, Counter, Info, Histogram, Summary
 except ImportError:
     raise ImportError('prometheus_client is not installed - Install it and try again.')
 from twisted.internet import reactor, task
 
 import nucypher
-from nucypher.blockchain.eth.agents import ContractAgency, StakingEscrowAgent
+from nucypher.blockchain.eth.agents import ContractAgency, StakingEscrowAgent, WorkLockAgent, PolicyManagerAgent
+from nucypher.blockchain.eth.actors import NucypherTokenActor
+from typing import List
+from nucypher.blockchain.eth.interfaces import BlockchainInterfaceFactory
 
-# Metrics
-known_nodes_guage = Gauge('known_nodes', 'Number of currently known nodes')
-work_orders_guage = Gauge('work_orders', 'Number of accepted work orders')
-missing_confirmation_guage = Gauge('missing_confirmations', 'Currently missed confirmations')
-learning_status = Enum('node_discovery', 'Learning loop status', states=['starting', 'running', 'stopped'])
-requests_counter = Counter('http_failures', 'HTTP Failures', ['method', 'endpoint'])
-host_info = Info('host_info', 'Description of info')
-active_stake_gauge = Gauge('active_stake', 'Active stake')
 
+class EventMetricsCollector:
 
-def collect_prometheus_metrics(ursula):
+    def __init__(self, staker_address, worker_address, contract_agent, event_name, argument_filters, metrics):
+        self.event_filter = contract_agent.contract.events[event_name].createFilter(fromBlock='latest',
+                                                                                    argument_filters=argument_filters)
+        self.metrics = metrics
+        self.event_name = event_name
+        self.staker_address = staker_address
+        self.worker_address = worker_address
+        self.contract_agent = contract_agent
+
+        if self.event_name == "ReStakeSet":
+            self.metrics["reStake"].set(self.contract_agent.is_restaking(staker_address))
+
+        if self.event_name == "WindDownSet":
+            self.metrics["windDown"].set(self.contract_agent.is_winding_down(staker_address))
+
+    def collect(self, node_metrics):
+        events = self.event_filter.get_new_entries()
+        for event in events:
+            for arg in self.metrics.keys():
+                if arg == "block_number":
+                    self.metrics["block_number"].set(event["blockNumber"])
+                    continue
+                self.metrics[arg].set(event['args'][arg])
+
+            if self.event_name == "WorkerSet":
+                node_metrics["current_worker_is_me_gauge"].set(
+                    self.contract_agent.get_worker_from_staker(self.staker_address) == self.worker_address)
+
+            if self.event_name == "Bid" or self.event_name == "Refund":
```

Comment:
I preface this by saying i'm unfamiliar with `prometheus` and `grafana` 😅 

Instead of having `event_collectors_config` dictionaries, and then having these special `event_name` cases ("WorkerSet", "Bid", "Refund") within the same `collect` function for the EventMetricsCollector, would it make sense to instead have a `BaseEventMetricsCollector` type class and then just have subclasses of that, each with their specific functionality? 

Currently what is done is:

1. define a config
2. create the collector from the config
3. include special uses cases in collect(...) function

Instead, the subclasses could be added to your event collectors list in `get_event_metrics_collectors ` directly. Their respective `collect()` functions eventually called which calls the base `collect` and each can add its own special functionality if required i.e. just instantiate a specific collector and add it to the `get_event_metrics_collectors` list.